src/Readers/ODataReader.v4/ODataCapabilitiesReader.cs

```diff
@@ -0,0 +1,196 @@
+using Microsoft.OData.Edm.Validation;
+using System;
+using System.Collections.Generic;
+using System.IO;
+using System.Linq;
+using System.Text;
+using System.Xml;
+using Vipr.Core;
+using Vipr.Core.CodeModel;
+
+namespace ODataReader.v4
+{
+    internal static class ODataCapabilitiesReader
+    {
+        private static IEdmModel s_capabilitiesModel;
+        private static readonly IEdmValueTerm s_insertRestrictionsTerm;
+        private static readonly IEdmValueTerm s_updateRestrictionsTerm;
+        private static readonly IEdmValueTerm s_deleteRestrictionsTerm;
+        private static readonly IEdmValueTerm s_expandRestrictionsTerm;
+
+        const string CapabilitiesInsertRestrictions = "Org.OData.Capabilities.V1.InsertRestrictions";
+        const string CapabilitiesUpdateRestrictions = "Org.OData.Capabilities.V1.UpdateRestrictions";
+        const string CapabilitiesDeleteRestrictions = "Org.OData.Capabilities.V1.DeleteRestrictions";
+        const string CapabilitiesExpandRestrictions = "Org.OData.Capabilities.V1.ExpandRestrictions";
+        
+        static ODataCapabilitiesReader()
+        {
+            using(var reader = new StringReader(Properties.Resources.CapabilitiesVocabularies))
+            {
+                IEnumerable<EdmError> errors;
+                if (!CsdlReader.TryParse(new[] { XmlReader.Create(reader) }, out s_capabilitiesModel, out errors))
+                {
+                    throw new InvalidOperationException("Could not load capabilities vocabulary from resources");
+                }
+            }
+
+            s_insertRestrictionsTerm = s_capabilitiesModel.FindDeclaredValueTerm(CapabilitiesInsertRestrictions);
+            s_updateRestrictionsTerm = s_capabilitiesModel.FindDeclaredValueTerm(CapabilitiesUpdateRestrictions);
+            s_deleteRestrictionsTerm = s_capabilitiesModel.FindDeclaredValueTerm(CapabilitiesDeleteRestrictions);
+            s_expandRestrictionsTerm = s_capabilitiesModel.FindDeclaredValueTerm(CapabilitiesExpandRestrictions);
+        }
+
+        internal static void GetCapabilitiesForEntitySet(OdcmProperty odcmProperty, IEdmModel edmModel, IEdmEntitySet entitySet)
+        {
+            if (odcmProperty == null) throw new ArgumentNullException("odcmProperty");
+            if (edmModel == null) throw new ArgumentNullException("edmModel");
+            if (entitySet == null) throw new ArgumentNullException("entitySet");
+
+            GetInsertRestrictions(odcmProperty, edmModel, entitySet);
+            GetUpdateRestrictions(odcmProperty, edmModel, entitySet);
+            GetDeleteRestrictions(odcmProperty, edmModel, entitySet);
+            GetExpandRestrictions(odcmProperty, edmModel, entitySet);
+        }
+
+        private static void GetInsertRestrictions(OdcmProperty odcmProperty, IEdmModel model, IEdmEntitySet entitySet)
+        {
+            IEdmValueAnnotation insertAnnotation = FindVocabularyAnnotation(model, entitySet, s_insertRestrictionsTerm);
```

Comment:
The name here is confusing because the method does not return a value. Consider modeling this as an extension method on odcmProperty, showing who is being acted upon, and rename it to something like SetNavigationProperty*Restrictions